Sources/SwiftSyntax/Syntax.swift

```diff
@@ -372,56 +330,171 @@ extension Syntax {
+  /// Underlying "green" node of this node
+  let raw: RawSyntax
+  /// Reference to the parent data. Or `nil` if this is the root.
+  let parent: SyntaxDataReference?
+  /// Index and position info in the tree.
+  let absoluteInfo: AbsoluteSyntaxInfo
+  /// Length of the children layout.
+  /// This is a cached information, equals to `raw.layoutView?.children.count ?? 0`.
+  let childCount: UInt32
+
+  // If the childCount > 0, the layout buffer (`SyntaxDataArenaReference? * childCount`) is tail allocated.
+}
+
+/// `SyntaxDataArena` manages the entire data of a "red" tree.
+final class SyntaxDataArena: @unchecked Sendable {
+  /// Mutex for locking the data when populating layout buffers.
+  private let mutex: PlatformMutex
+
+  /// Allocator.
+  private let allocator: BumpPtrAllocator
+
+  /// Retaining reference to the arena of the _root_ ``RawSyntax``
+  private let rawArena: RetainedSyntaxArena
+
+  /// Root node.
+  let root: SyntaxDataReference
+
+  init(raw: RawSyntax, rawNodeArena: RetainedSyntaxArena) {
+    precondition(rawNodeArena == raw.arenaReference)
+    assert(MemoryLayout<SyntaxData>.alignment >= MemoryLayout<UnsafePointer<SyntaxData>?>.alignment)
+
+    self.mutex = PlatformMutex.create()
+    self.allocator = BumpPtrAllocator(initialSlabSize: Self.slabSize(for: raw))
+    self.rawArena = rawNodeArena
+    self.root = Self.createDataImpl(allocator: allocator, raw: raw, parent: nil, absoluteInfo: .forRoot(raw))
   }
 
-  /// Extract the Syntax value.
-  @inline(__always)
-  var value: Syntax {
-    Syntax(raw, info: info.takeUnretainedValue())
+  deinit {
+    // print("nodeCount: \(root.pointee.raw.totalNodes), estimated: \(Self.slabSize(for: root.pointee.raw)), allocated: \(allocator.totalByteSizeAllocated)")
+    self.mutex.destroy()
   }
 
-  /// Get the parent of the Syntax value, but without retaining it.
-  @inline(__always)
-  var parent: UnownedSyntax? {
-    return info._withUnsafeGuaranteedRef {
-      switch $0.info.unsafelyUnwrapped {
-      case .nonRoot(let info):
-        return UnownedSyntax(info.parent)
-      case .root(_):
-        return nil
-      }
+  /// Return the childen data of the given node.
+  ///
+  /// The layout buffer is created and cached when it's first accssed.
+  func layout(for parent: SyntaxDataReference) -> SyntaxDataReferenceBuffer {
+    let childCount = Int(truncatingIfNeeded: parent.pointee.childCount)
+
+    // Return empty buffer for the node with no children.
+    guard childCount != 0 else {
+      return SyntaxDataReferenceBuffer()
     }
+
+    // The storage of the buffer address is allocated next to the SyntaxData.
+    let baseAddressRef = parent.advanced(by: 1)
+      .unsafeRawPointer
+      .assumingMemoryBound(to: UnsafePointer<SyntaxDataReference?>?.self)
+
+    // If the buffer is already populated, return it.
+    if let baseAddress = baseAddressRef.pointee {
+      return SyntaxDataReferenceBuffer(UnsafeBufferPointer(start: baseAddress, count: childCount))
```

Comment:
Probably we should not read the memory before the lock, or we should use atomic read/write. But does it really matter?